src/rgw/driver/sfs/types.cc

```diff
@@ -439,8 +447,11 @@ void Bucket::_delete_object_version(
   version.delete_time = now;
   version.mtime = now;
   version.object_state = ObjectState::DELETED;
-  db_versioned_objs.store_versioned_object(version);
+  const bool ret = db_versioned_objs.store_versioned_object_if_state(
+      version, {ObjectState::OPEN, ObjectState::COMMITTED}
```

Comment:
If get only returns objects with state = `COMMITTED`... how is it possible to delete an OPEN object?

`Object::delete_obj` calls `refresh_meta` --> `bucket->get()` --> `try_fetch_from_database` --> `get_committed_versioned_object`

The result of `bucket::get` is `throw UnknownObjectException();` and nothing is deleted.

For versioned buckets I suspect this race condition could end up creating a delete_marker and when the upload (the one OPEN) is done... it will commit the version with higher commit time, so it would prevail on top of the delete marker (because we get the last version based on the commit time)

This is the code that would end up creating the delete marker: (see comments with ***)
```c++
int SFSObject::SFSDeleteOp::delete_obj(
    const DoutPrefixProvider* dpp, optional_yield y
) {
  lsfs_dout(dpp, 10) << "bucket: " << source->bucket->get_name()
                     << " bucket versioning: "
                     << source->bucket->versioning_enabled()
                     << ", object: " << source->get_name()
                     << ", instance: " << source->get_instance() << dendl;

  // do the quick and dirty thing for now
  ceph_assert(bucketref);
  if (!source->objref) {
    source->refresh_meta(); // *** HERE WE REFRESH META, CALLS Bucket::get and returns nothing (exception) 
  }

  auto version_id = source->get_instance();
  std::string delete_marker_version_id;
  if (source->objref) { // *** source->objref is nullptr (Bucket::get returned nothing because of the exception)
    bucketref->delete_object(
        source->objref, source->get_key(), source->bucket->versioning_enabled(),
        delete_marker_version_id
    );
 // *** THIS WILL BE CALLED IF THE USER WAS DELETING THE OBJECT WITHOUT SPECIFYING A VERSION
 // *** A DELETE MARKER WILL BE CREATED
  } else if (source->bucket->versioning_enabled() && source->get_instance().empty()) {
    // create delete marker
    // even the object does not exist AWS creates a delete marker for it
    // if versioning is enabled and a specific version was not specified
    version_id =
        bucketref->create_non_existing_object_delete_marker(source->get_key());
  }
```

I agree with returning only COMMITTED objects (not open), but I'm not so sure about deleting OPEN objects.

I think we should differentiate when the object is OPEN, because in a race condition like the one described for versioned buckets we could end up with a non desired state.
If `Bucket::get` returns the OPEN object we can then differentiate and avoid the creation of that delete marker. 

I agree for reads only COMMITTED objects should be returned.

The problem is that S3 expects us to create that delete marker when deleting an object that does not exists.
There is an `s3test` for it.

I'm not sure what should we do in a race condition between UPLOAD vs DELETE and if we create the delete marker before... maybe we should delete it when the object is finally committed. 
I think it makes no sense to have a visible object that can also has a delete marker....

This could be a good topic to discuss on a dev call on Wednesdays. :)

Reply:
I think we should not do the delete marker decision on OPEN versions.

OPEN versions are not guaranteed to transition to COMMITTED or DELETED. They may never leave OPEN when a client aborts a PUT for instance. Or we crash. 

While we store versions and delete markers in the same table with the same attributes, I think we must treat the *_time separate:
No:  t_v1 < t_v2 < t_deletemarker1 < t_v3 < t_deletemarker2
Yes: t_v1 < t_v2 < t_v3 < t_deletemarker1 < t_deletemarker2

My read of the spec is that delete markers are always on top of versions, regardless when they were set.